core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java

```diff
@@ -0,0 +1,445 @@
+ */
+public class RelCopyOnWriteVisitor implements RelVisitor<Rel, RuntimeException> {
+  @Override
+  public Aggregate visit(Aggregate aggregate) {
+    Rel oldInput = aggregate.getInput();
+    Rel newInput = oldInput.accept(this);
+    return oldInput == newInput
+        ? aggregate
+        : ImmutableAggregate.builder().from(aggregate).input(newInput).build();
+  }
+
+  @Override
+  public EmptyScan visit(EmptyScan emptyScan) {
+    return emptyScan;
+  }
+
+  @Override
+  public Fetch visit(Fetch fetch) {
+    Rel oldInput = fetch.getInput();
+    Rel newInput = oldInput.accept(this);
+    return oldInput == newInput
+        ? fetch
+        : ImmutableFetch.builder().from(fetch).input(newInput).build();
+  }
+
+  @Override
+  public Filter visit(Filter filter) {
+    Rel oldInput = filter.getInput();
+    Rel newInput = oldInput.accept(this);
+    Expression oldCondition = filter.getCondition();
+    Expression newCondition = visitExpression(oldCondition);
+    return oldInput == newInput && oldCondition == newCondition
+        ? filter
+        : ImmutableFilter.builder().from(filter).input(newInput).condition(newCondition).build();
+  }
+
+  @Override
+  public Join visit(Join join) {
+    Rel oldLeft = join.getLeft();
+    Rel newLeft = oldLeft.accept(this);
+    Rel oldRight = join.getRight();
+    Rel newRight = oldRight.accept(this);
+    Expression oldCondition = null, newCondition = null;
+    if (join.getCondition().isPresent()) {
+      newCondition = visitExpression(oldCondition = join.getCondition().get());
+    }
+    Expression oldFilter = null, newFilter = null;
+    if (join.getPostJoinFilter().isPresent()) {
+      newFilter = visitExpression(oldFilter = join.getPostJoinFilter().get());
+    }
+    if (oldLeft == newLeft
+        && oldRight == newRight
+        && oldCondition == newCondition
+        && oldFilter == newFilter) {
+      return join;
+    }
+    ImmutableJoin.Builder builder =
+        ImmutableJoin.builder().from(join).left(newLeft).right(newRight);
+    if (join.getCondition().isPresent()) {
+      builder.condition(newCondition);
+    }
+    if (join.getPostJoinFilter().isPresent()) {
+      builder.postJoinFilter(newFilter);
+    }
+    return builder.build();
+  }
+
+  @Override
+  public NamedScan visit(NamedScan namedScan) {
+    return namedScan;
+  }
+
+  private List<Expression> copyOnWrite(List<Expression> oldExpressions) {
+    int expressionCount = oldExpressions.size();
+    List<Expression> newExpressions = oldExpressions;
+    for (int i = 0; i < expressionCount; i++) {
+      Expression oldExpr = oldExpressions.get(i);
+      Expression newExpr = visitExpression(oldExpr);
+      if (oldExpr != newExpr) {
+        if (oldExpressions == newExpressions) {
+          newExpressions = new CopyOnWriteArrayList<>(oldExpressions);
+        }
+        newExpressions.set(i, newExpr);
+      }
+    }
+    return newExpressions;
+  }
+
+  @Override
+  public Project visit(Project project) {
```

Comment:
rewrite using above functions
```
public Project visit(Project project) {
  var newInput = project.getInput().accept(this);
  var newExpressions = copyOnWrite(project.getExpressions());
  if (!newInput.isPresent() && !newExpressions.isPresent()) {
    return Optional.empty();
  }
  return Optional.of(ImmutableProject.builder().from(project)
    .input(newInput.orElse(project.getInput())
    .expressions(newExpressions.orElse(project.getExpressions())
    .build());
}
```

Note that this isn't substantially shorter but it makes things explicit.